server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java

```diff
@@ -334,32 +334,88 @@ protected FieldMapper clone() {
     }
 
     @Override
-    public FieldMapper merge(Mapper mergeWith) {
+    public final FieldMapper merge(Mapper mergeWith) {
         FieldMapper merged = clone();
-        merged.doMerge(mergeWith);
+        List<String> conflicts = new ArrayList<>();
+        if (mergeWith instanceof FieldMapper == false) {
+            throw new IllegalArgumentException("mapper [" + fieldType.name() + "] cannot be changed from type ["
+            + contentType() + "] to [" + mergeWith.getClass().getSimpleName() + "]");
+        }
+        merged.mergeSharedOptions((FieldMapper)mergeWith, conflicts);
```

Comment:
Small comment, calling `mergeOptions` directly from here could be nice:
```
FieldMapper fieldMapper = (FieldMapper) mergeWith;
merged.mergeSharedOptions(fieldMapper, conflicts);
merged.mergeOptions(fieldMapper, conflicts);
```

Reply:
++, I've rearranged things here.